### PR TITLE
Extend functionality of zqd logging facilities

### PR DESCRIPTION
**Pull request by mattnibs**
Adds several features to zqd logging:
1. Adds either client provided or generated unique request ID to each
http request and appends request ID as a field to request-specific
generated logs.
2. Add name-based filter to loggers
3. Add functionality to specify multiple output paths for logging
via a zqd config. Each path can specify a different file mode:
Append (appends to existing log files), Truncate, and Rotate (which
uses lumberjack.Logger to rotate logs).
4. Add access log to http requests.

### CODE REVIEW

File: zqd/handlers_test.go [hunk 1]
```diff
@@ -232,14 +263,21 @@ func execSearch(t *testing.T, c *zqd.Core, space, prog string) string {
 }
 
 func createTempDir(t *testing.T) string {
-	dir, err := ioutil.TempDir("", t.Name())
+	// Replace filepath.Separator with '-' so it doesn't try to access dirs that
+	// don't exist.
+	name := strings.Replace(t.Name(), string(filepath.Separator), "-", -1)
```
Comment (nwt) [inline zqd/handlers_test.go]:
```suggestion
	name := strings.ReplaceAll(t.Name(), string(filepath.Separator), "-")
```

File: cmd/zqd/listen/command.go [hunk 2]
```diff
@@ -23,28 +26,70 @@ The listen command launches a process to listen on the provided interface and
 	New: New,
 }
 
+// Example configfile
+// loggers:
+// - path: ./data/access.log
+//   name: "http.access"
+//   level: info
+//   mode: truncate
+
```
Comment (nwt) [inline cmd/zqd/listen/command.go]: Maybe have this adjacent to loadConfigFile so the two are more likely to stay in sync.

File: cmd/zqd/logger/file.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	// restarts.
+	FileModeTruncate = "truncate"
+	// FileModeRotate will enable log rotation for log files.
+	FileModeRotate = "rotate"
+)
+
+func OpenFile(path string, mode FileMode) (zapcore.WriteSyncer, error) {
+	switch path {
+	case "stdin":
+		return zapcore.Lock(os.Stdin), nil
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply cmd/zqd/logger/file.go]: whoops, thanks

File: cmd/zqd/logger/file.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+func logrotate(path string, mode FileMode) (zapcore.WriteSyncer, error) {
+	// Make sure directory exists
+	dir, _ := filepath.Split(path)
+	info, err := os.Stat(dir)
+	if err != nil {
+		return nil, err
+	}
+	if !info.IsDir() {
+		return nil, errors.New("expected path directories to already exist")
+	}
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply cmd/zqd/logger/file.go]: 👍

File: cmd/zqd/listen/command.go (hunk 2 above)
Comment (mattnibs) [reply cmd/zqd/listen/command.go]: Good idea.

**Comment by alfred-landrum** [comment]
@mattnibs : would you provide some example log output from using this with the brim app, please?

**Comment by mattnibs** [comment]
@alfred-landrum sure. Running brim against zqd with this config:
```
loggers:
- path: ./data/access.log
  name: "http.access"
  level: info
  mode: rotate
- path: ./data/zqd.log
  name: "zqd"
  level: info
  mode: rotate
```
I get in `access.log` (abbreviated):
```
{"level":"info","ts":1584379462.695507,"logger":"http.access","msg":"Request completed","request_id":"1","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"[::1]:59207","request_content_length":1003,"url":"/search?format=zjson","elapsed":0.000253663,"response_content_length":21,"status_code":404}
{"level":"info","ts":1584379502.1226501,"logger":"http.access","msg":"Request completed","request_id":"2","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"[::1]:59207","request_content_length":230,"url":"/search?format=zjson","elapsed":0.000071302,"response_content_length":21,"status_code":404}
{"level":"info","ts":1584379502.125317,"logger":"http.access","msg":"Request completed","request_id":"3","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"[::1]:59207","request_content_length":351,"url":"/search?format=zjson","elapsed":0.000072759,"response_content_length":21,"status_code":404}
{"level":"info","ts":1584379542.4489012,"logger":"http.access","msg":"Request completed","request_id":"4","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"[::1]:59207","request_content_length":0,"url":"/space","elapsed":0.001329315,"response_content_length":3,"status_code":200}
{"level":"info","ts":1584379542.487246,"logger":"http.access","msg":"Request completed","request_id":"5","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"[::1]:59207","request_content_length":0,"url":"/space","elapsed":0.000435085,"response_content_length":3,"status_code":200}
{"level":"info","ts":1584379558.7751412,"logger":"http.access","msg":"Request completed","request_id":"6","host":"localhost:9867","method":"POST","proto":"HTTP/1.1","remote_addr":"[::1]:59207","request_content_length":52,"url":"/space","elapsed":0.0004188,"response_content_length":76,"status_code":200}
{"level":"info","ts":1584379562.2027972,"logger":"http.access","msg":"Request completed","request_id":"8","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"[::1]:59217","request_content_length":0,"url":"/space/all.pcap.brim","elapsed":0.000178415,"response_content_length":121,"status_code":200}
{"level":"info","ts":1584379563.205507,"logger":"http.access","msg":"Request completed","request_id":"9","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"[::1]:59217","request_content_length":0,"url":"/space/all.pcap.brim","elapsed":0.000155445,"response_content_length":121,"status_code":200}
{"level":"info","ts":1584379563.292762,"logger":"http.access","msg":"Request completed","request_id":"10","host":"localhost:9867","method":"GET","proto":"HTTP/1.1","remote_addr":"[::1]:59217","request_content_length":0,"url":"/space/all.pcap.brim","elapsed":0.000163234,"response_content_length":216,"status_code":200}
```
zqd.log:
```
{"level":"info","ts":1584379438.952985,"logger":"zqd","msg":"Listening","addr":"[::]:9867","datadir":"/Users/nibs/Code/zq","pprof_routes":false}
```

File: zqd/handlers.go [hunk 5]
```diff
@@ -245,9 +243,7 @@ func handlePacketPost(c *Core, w http.ResponseWriter, r *http.Request) {
 	taskID := c.getTaskID()
 	taskStart := api.TaskStart{Type: "TaskStart", TaskID: taskID}
 	if err = pipe.Send(taskStart); err != nil {
-		// Probably an error writing to socket, log error.
-		// XXX This should be zap instead.
-		log.Print(err)
+		logger.Warn("Err sending payload", zap.Error(err))
```
Comment (nwt) [inline zqd/handlers.go]:
```suggestion
		logger.Warn("Error sending payload", zap.Error(err))
```

File: cmd/zqd/listen/command.go [hunk 6]
```diff
@@ -23,28 +26,72 @@ The listen command launches a process to listen on the provided interface and
...
 }
 
+// defaultLogger ignores output from the access logger.
+var defaultLogger = []logger.Config{
+	{
+		Name:  "zqd",
+		Path:  "stderr",
+		Level: zap.InfoLevel,
+	},
+}
```
Comment (mattnibs) [inline cmd/zqd/listen/command.go]: @alfred-landrum as we discussed elsewhere, I added a default logger which trims out the noisy access logs.

**Pull request merged**